allauth/socialaccount/providers/apple/views.py

```diff
@@ -76,54 +77,91 @@ def get_verified_identity_data(self, id_token):
             raise OAuth2Error("Invalid id_token") from e
 
     def parse_token(self, data):
-        access_token_data = data["access_token_data"]
-        identity_data = data["identity_data"]
+        identity_data = self.get_verified_identity_data(data["id_token"])
 
-        expires_in = access_token_data[self.expires_in_key]
+        expires_in = data[self.expires_in_key]
         expires_at = timezone.now() + timedelta(seconds=int(expires_in))
 
         token = SocialToken(
-            token=access_token_data["access_token"],
-            token_secret=access_token_data["refresh_token"],
+            token=data["access_token"],
+            token_secret=data["refresh_token"],
             expires_at=expires_at,
         )
-        token.user_data = identity_data
+        # `user_data` is a big flat dictionary with the parsed JWT claims
+        # access_tokens, and user info from the apple post.
+        token.user_data = {**data, **identity_data}
         return token
 
     def complete_login(self, request, app, token, **kwargs):
         extra_data = token.user_data
         login = self.get_provider().sociallogin_from_response(request=request, response=extra_data)
         login.state["id_token"] = token.user_data
+
+        # We can safely remove the apple login session now
+        # Note: The cookie will remain, but it's set to delete on browser close
+        request.apple_login_session.delete()
+
         return login
 
     def get_user_scope_data(self, request):
-        user_scope_data = get_request_param(request, "user", "")
+        user_scope_data = request.apple_login_session.get("user", "")
         try:
             return json.loads(user_scope_data)
         except json.JSONDecodeError:
             # We do not care much about user scope data as it maybe blank
             # so return blank dictionary instead
             return {}
 
-    def get_identity_data(self, request):
-        id_token = get_request_param(request, "id_token")
-        identity_data = self.get_verified_identity_data(id_token=id_token)
-        identity_data["user_scope_data"] = self.get_user_scope_data(request)
-        return identity_data
-
     def get_access_token_data(self, request, app, client):
-        # Parse id_token and other form_post response data
-        identity_data = self.get_identity_data(request)
+        """ We need to gather the info from the apple specific login """
+        add_apple_session(request)
 
         # Exchange `code`
         code = get_request_param(request, "code")
         access_token_data = client.get_access_token(code)
 
         return {
-            "identity_data": identity_data,
-            "access_token_data": access_token_data
+            **access_token_data,
+            **self.get_user_scope_data(request),
+            "id_token":request.apple_login_session.get("id_token")
         }
```

Comment:
Keeping this flat means that the shape of the `data` argument to `parse_token` is much more similar to other oauth providers. Additionally it means that it's possible to use `parse_token` with data which is sent from an iOS app (i.e. data from https://developer.apple.com/documentation/authenticationservices/asauthorizationappleidcredential)